hazelcast/src/main/java/com/hazelcast/spi/discovery/DiscoveryStrategy.java

```diff
@@ -74,7 +91,10 @@
      * in case of the default implementation is to be used
      * @since 3.7
      */
-    PartitionGroupStrategy getPartitionGroupStrategy();
+    @Deprecated
+    default PartitionGroupStrategy getPartitionGroupStrategy() {
```

Comment:
what was the reason that we don't return `null` in this method also? i think, not to violate least astonishment principle, we can also return `null` here.

Reply:
I am coming from a principle to make the biggest bang for the least amount of code.
This way, the implementor doesn't have to override `getPartitionGroupStrategy()` whenever they implement `DiscoveryStrategy` and possibly can use functional interfaces to discover nodes.
The reason original code returns `null` is because that was the best way to design API
under Java 6/7 but with Java 8 and default methods and Lambdas, the more automatic way
works much better and doesn't break compatibility.

I guess you are coming from a principle of "least changes" I understand but I think this way is better.

Reply:
I'm not in favor of changing these in patch release. For 5.0 we can do it as you suggest but for 4.2 patch release, both `getPartitionGroupStrategy` and `getPartitionGroupStrategy(Collection<? extends Member> allMembers)` can return `null` to indicate default implementation is to be used.

Reply:
ok